core/src/main/java/com/linecorp/armeria/common/encoding/AbstractStreamDecoder.java

```diff
@@ -80,4 +105,25 @@ protected HttpData fetchDecoderOutput() {
 
         return HttpData.wrap(decoded);
     }
+
+    private void maybeCheckOverflow(@Nullable ByteBuf decoded, ByteBuf newBuf) {
+        if (!checkOverflow) {
+            return;
+        }
+
+        long length = newBuf.readableBytes();
+        if (decoded != null) {
+            length += decoded.readableBytes();
+        }
+        if (length > maxLength) {
+            if (decoded != null) {
+                decoded.release();
+            }
+            newBuf.release();
+            throw ContentTooLargeException.builder()
```

Comment:
Could you check the call path of this method whether we throw this exception where we shouldn't?
It seems like `AbstractStreamDecoder.finish()` is called in `onComplete` and the exception isn't caught.

Reply:
Nice point. We need to call `delegate.onError()` if `beforeError()` raises an exception. 
Let me check possible cases.

Reply:
Fixed to handle `ContentTooLargeException` in:
- `beforeComplete()`: Halt `onComplete()` and call `onError()` instead
- `beforeError()`: Create a `CompositeException` and set both the original exception and `ContentTooLargeException`
- `onCancellation()`: Since the stream was canceled, just warn the error message.